lib/cosmos/interfaces/interface.rb

```diff
@@ -220,6 +306,56 @@ def set_option(option_name, option_values)
     def post_identify_packet(packet)
     end
 
-  end # class Interface
+    # Called to perform modifications on a command packet before it is turned
+    # into packet data to send. An example would be to override a known packet
+    # value. After this method is called the pre_write_data method is called.
+    #
+    # @param packet [Packet] Original packet
+    # @return [Packet] Potentially modified packet
+    def pre_write_packet(packet)
+      packet
+    end
+
+    # Called to write a packet to the underlying interface. Subclasses should
+    # implement this method if they need to do something with the packet during
+    # the write. Otherwise simply implement write_data.
+    #
+    # @param packet [Packet] Packet data
+    # @return [Packet] The original packet
+    def write_packet(packet)
+      packet
+    end
 
-end # module Cosmos
+    # Called to perform modifications on write data before writing it over the
+    # interface.
+    # TODO: Example of using this interface
+    # After this method is called the post_write_data method is called.
+    #
+    # @param packet_data [String] Raw packet data
+    # @return [String] Potentially modified packet data
+    def pre_write_data(packet_data)
+      @raw_logger_pair.write_logger.write(packet_data)
```

Comment:
There is no @stream at the generic interface level. When I looked at the existing protocols none of them actually modify the stream thus I assumed raw logging can occur at a higher level. It seems like it can (should) be done in the lowest level read_data and write_data methods (which subclasses must implement). I'll move it there and see if it makes sense with the inheritance structure.